components/supervisor-api/notification.proto

```diff
@@ -11,60 +11,129 @@ import "google/api/annotations.proto";
 option go_package = "github.com/gitpod-io/gitpod/supervisor/api";
 option java_package = "io.gitpod.supervisor.api";
 
-// Notification serivce allows external processes to notify the user and ask for decisions.
+// Notification serivce allows external processes to notify the user and ask for
+// decisions.
 service NotificationService {
 
-    // Prompts the user and asks for a decision. Typically called by some external process.
-    // If the list of actions is empty this service returns immediately,
-    // otherwise it blocks until the user has made their choice.
-    rpc Notify(NotifyRequest) returns (NotifyResponse) {
-        option (google.api.http) = {
-            post: "/v1/notification/notify"
-        };
-    }
-
-    // Subscribe to notifications. Typically called by the IDE.
-    rpc Subscribe(SubscribeRequest) returns (stream SubscribeResponse) {
-        option (google.api.http) = {
-            get: "/v1/notification/subscribe"
-        };
-    }
-
-    // Report a user's choice as a response to a notification. Typically called by the IDE.
-    rpc Respond(RespondRequest) returns (RespondResponse) {
-        option (google.api.http) = {
-            post: "/v1/notification/respond"
-        };
-    }
+  // Prompts the user and asks for a decision. Typically called by some external
+  // process. If the list of actions is empty this service returns immediately,
+  // otherwise it blocks until the user has made their choice.
+  rpc Notify(NotifyRequest) returns (NotifyResponse) {
+    option (google.api.http) = {
+      post : "/v1/notification/notify"
+    };
+  }
+
+  // Subscribe to notifications. Typically called by the IDE.
+  rpc Subscribe(SubscribeRequest) returns (stream SubscribeResponse) {
+    option (google.api.http) = {
+      get : "/v1/notification/subscribe"
+    };
+  }
+
+  // Report a user's choice as a response to a notification. Typically called by
+  // the IDE.
+  rpc Respond(RespondRequest) returns (RespondResponse) {
+    option (google.api.http) = {
+      post : "/v1/notification/respond"
+    };
+  }
+
+  // SetActiveClient to inform supervisor about
+  // which is the latest client actively used by the user.
+  // Typically called by the IDE.
+  // Will response what client should do like `open` `preview`
+  rpc SetActiveClient(SetActiveClientRequest)
+      returns (stream SetActiveClientResponse) {
```

Comment:
why is it stream?

Reply:
We use stream because we want to ask client to do action here, so same as `Subscribe`.

If we choose to extend `NotifyRequest` we don't need this `stream` here

Reply:
yeah, the idea behind `SetActiveClient` is for the client to "subscribe" as active client and by returning a stream we can then later (when an Action request arrives), ask the active client to perform the action.

Reply:
ok, not i'm starting getting it, so you we could have 2 new APIs like `SubscribeActive` and `NotifyActive`? The last caller of `SubscribeActive` wins and close currently active? If the client becomes inactive it simply drops `SubscribeActive`?